### PR TITLE
Adds ability to trigger events from commit status changes

### PR DESCRIPTION
**Pull request by gandelman-a**
This adds support for triggering on github status updates.

The status matching logic was moved out of the ChangeishFilter
and into BaseFilter so it can also be used by EventFilters.  Config
schema for the github trigger has been updated to accept a list of
statuses, in the "github_user:context:status" format.

Signed-off-by: Adam Gandelman <adamg@ubuntu.com>

### CODE REVIEW

File: tests/base.py [hunk 1]
```diff
@@ -798,6 +813,14 @@ def getPull(self, owner, project, number):
         }
         return data
 
+    def getPullBySha(self, sha):
+        prs = list(set([p for p in self.pull_requests if sha == p.head_sha]))
+        if len(prs) > 1:
+            raise Exception('Multiple pulls found with head sha: %s' % sha)
```
Comment (gandelman-a) [inline tests/base.py:818]: Im not sure if we want to raise an error here or what? Is this case even possible?

**Comment by anne-bonny** [comment]
Build failed.

- http://logs.bonnyci.com/check_github/BonnyCI/zuul/12/1485466534.61/tox_py27 : FAILURE in 19m 33s

File: tests/base.py (hunk 1 above)
Comment (ConsoleCatzirl) [reply tests/base.py:818]: Yeah, this is possible, it would mean that there are two identical PRs (or rather two PRs with differently-named but identical branches). Definitely a strange and unlikely corner case, but since we don't know which is correct I think an exception is correct.

File: zuul/connection/github.py [hunk 2]
```diff
@@ -440,6 +455,23 @@ def getPull(self, owner, project, number):
         log_rate_limit(self.log, self.github)
         return pr
 
+    def getPullBySha(self, sha):
+        query = '%s type:pr is:open' % sha
+        pulls = set()
+        for issue in self.github.search_issues(query=query):
+            pr_url = issue.pull_request.get('url')
```
Comment (ConsoleCatzirl) [inline zuul/connection/github.py:461]: Do all issue objects have a pull_request attribute, or do we need to check for an AttributeError here?

File: zuul/connection/github.py [hunk 3]
```diff
@@ -185,6 +185,21 @@ def _event_pull_request_review(self, request):
...
+        action = body.get('action')
+        if action == 'pending':
+            return
+        pr_body = self.connection.getPullBySha(body['sha'])
+        if pr_body is None:
+            return
+
+        event = self._pull_request_to_event(pr_body)
+        event.account = self._get_sender(body)
+        event.comment = body.get('comment', {}).get('body')
```
Comment (ConsoleCatzirl) [inline zuul/connection/github.py]: Do status events ever come with a comment? I don't see that in the [example payload](https://developer.github.com/v3/activity/events/types/#statusevent), and I'm not sure what this could be.

File: zuul/connection/github.py (hunk 2 above)
Comment (gandelman-a) [reply zuul/connection/github.py:461]: AFAICS they *should*, we're searching by 'type:pr' so anything returned should have a 'pull_request' key in its json body, the contents of which is what the attribute access is fetching here.

File: zuul/connection/github.py (hunk 3 above)
Comment (gandelman-a) [reply zuul/connection/github.py]: Good call, removed.

**Review by ConsoleCatzirl — APPROVED** [review]
lgtm

File: tests/base.py [hunk 4]
```diff
@@ -713,6 +713,21 @@ def _getPullRequestEvent(self, action):
         }
         return (name, data)
 
+    def getCommitStatusEvent(self, context, state='success', user='zuul'):
+        name = 'status'
+        data = {
+            'state': state,
+            'sha': self.head_sha,
+            'description': 'Test results for %s: %s' % (self.head_sha, state),
+            'target_url': 'http://bonnyci.org/%s' % self.head_sha,
```
Comment (omgjlk) [inline tests/base.py]: Having bonnyci in there is probably inappropriate. This should be more generic.

File: tests/base.py [hunk 5]
```diff
@@ -798,6 +813,14 @@ def getPull(self, owner, project, number):
         }
         return data
 
+    def getPullBySha(self, sha):
+        prs = list(set([p for p in self.pull_requests if sha == p.head_sha]))
```
Comment (omgjlk) [inline tests/base.py:816]: Why a list of a set of a list (comprehension)?

File: tests/test_github.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2015 GoodData
```
Comment (omgjlk) [inline tests/test_github.py]: Was removing the copyright intentional?

File: tests/test_github_requirements.py [hunk 7]
```diff
@@ -64,6 +55,26 @@ def _test_require_status(self, project, job):
...
+        self.sched.reconfigure(self.config)
+        self.registerJobs()
+
+        A = self.fake_github.openFakePullRequest('org/project2', 'master', 'A')
+
+        # An error status should not cause it to be enqueued
+        A.setStatus(A.head_sha, 'error', 'null', 'null', 'check')
+        self.fake_github.emitEvent(A.getCommitStatusEvent('error'))
+        self.waitUntilSettled()
+        self.assertEqual(len(self.history), 0)
```
Comment (omgjlk) [inline tests/test_github_requirements.py:73]: We should add a test here that puts in a success status from somebody other than zuul, to make sure that we're really matching the whole user+context+state set.

File: zuul/connection/github.py [hunk 8]
```diff
@@ -440,6 +454,23 @@ def getPull(self, owner, project, number):
...
+                continue
+            # the issue provides no good description of the project :\
+            owner, project, _, number = pr_url.split('/')[4:]
+            pr = self.github.pull_request(owner, project, number)
+            if pr.head.sha != sha:
+                continue
+            pulls.add(pr.as_dict())
+        if len(pulls) > 1:
+            raise Exception('Multiple pulls found with head sha %s' % sha)
+        return pulls.pop()
```
Comment (omgjlk) [inline zuul/connection/github.py]: Before we return, this needs a call to update the API rate limits.

`log_rate_limit(self.log, self.github)`

File: tests/fixtures/layout-github-requirement-status.yaml [hunk 9]
```diff
@@ -12,25 +12,25 @@ pipelines:
...
-#    failure:
-#      github:
-#        status: true
+  - name: trigger
+    manager: IndependentPipelineManager
+    source: github
+    trigger:
+      github:
+        - event: status
+          statuses:
```
Comment (omgjlk) [inline tests/fixtures/layout-github-requirement-status.yaml]: With other layout terms, the term is singular, not plural.

File: zuul/trigger/github.py [hunk 10]
```diff
@@ -63,13 +64,15 @@ def toList(x):
...
                      'push',
+                     'status',
                      'tag',
                      )),
         'branch': toList(str),
         'ref': toList(str),
         'comment': toList(str),
         'label': toList(str),
         'state': toList(str),
+        'statuses': toList(str)
```
Comment (omgjlk) [inline zuul/trigger/github.py]: I think this should be `status` not plural.

File: zuul/connection/github.py [hunk 11]
```diff
@@ -185,6 +185,20 @@ def _event_pull_request_review(self, request):
         event.type = 'pr-review'
         return event
 
+    def _event_status(self, request):
+        body = request.json_body
+        action = body.get('action')
+        if action == 'pending':
+            return
```
Comment (omgjlk) [inline zuul/connection/github.py:192]: Seems like a hard decision here to not allow triggering on a state of `pending`. But we report up the state of either `success`. Shouldn't we also report `pending` and let layout authors decide if they want to do anything on pending or not?

**Review by omgjlk — CHANGES_REQUESTED** [review]
Docs need to be updated to reflect the ability to trigger on status, and what filtering is available.

File: tests/base.py (hunk 5 above)
Comment (gandelman-a) [reply tests/base.py:816]: Quick and easy way to remove duplicates from a list

File: tests/test_github.py (hunk 6 above)
Comment (gandelman-a) [reply tests/test_github.py]: Hmm, oops

File: zuul/connection/github.py (hunk 11 above)
Comment (gandelman-a) [reply zuul/connection/github.py:192]: We could, tho I had our use case described in the flowchart at [1] in mind while doing this.  We could certainly let 'pending' events through but that'd result in some expensive API calls later on in order to make the scheduling decision.  

[1] https://github.com/BonnyCI/lore/blob/master/developers/contributing/designs/github-pr-check-gate.md

Comment (omgjlk) [reply zuul/connection/github.py:192]: Ah. Our design document was mostly about how we'll set up the layout for gates and such, I did' mean for it to narrowly focus what the underlying Zuul code will support.

That said, I get what you're saying about API calls on a pending status. I can accept having it this way in 2.5, and maybe even v3, but we'll let upstream debate that on v3.

File: tests/test_github_requirements.py [hunk 12]
```diff
@@ -64,6 +55,33 @@ def _test_require_status(self, project, job):
...
+
+        # An success status from unknown user should not cause it to be
+        # enqueued
+        A.setStatus(A.head_sha, 'error', 'null', 'null', 'check', user='foo')
```
Comment (omgjlk) [inline tests/test_github_requirements.py:77]: Comment says success, but function call says error.

**Review by omgjlk — APPROVED** [review]

**Comment by anne-bonny** [comment]
Starting gate_github jobs.
